### PR TITLE
refactor(schemas): Pull out as `cargo-util-schemas`

### LINKED ISSUE #12801: Official API for parsing `Cargo.toml`s schema
### Problem

Something is needed by crates.io for creating index entries, rather than reading the json blob which might be out of sync.  They switched to community developed crates but this creates hidden relationships when making manifest changes (rust-lang/crates.io#6914).

### Proposed Solution

Bigger requests, like #4614, are blocked on the fact that the code ties into everything within cargo.  So instead, let's get an incremental step by just putting out the schemas, with no business logic.

### Notes

Steps
- [x] #12654
- [x] #12768
- [x] #12881
- [x] #12902
- [x] #12911
- [x] #12926
- [x] #12940
- [x] #12948
- [x] #12954
- [x] #12960
- [x] #12971
- [x] #13000
- [x] #13080
- [x] #13097
- [x] #13128
- [x] #13154
- [x]  #13164
- [x] #13166
- [ ] #13178

Related work
- #12924

See also #2259

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

Fixes #12801

### How should we test and review this PR?

See the individual commits for further justifications on the changes

### CODE REVIEW

File: crates/cargo-util-schemas/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "cargo-util-schemas"
+version = "0.1.0"
+rust-version.workspace = true
+edition.workspace = true
+license.workspace = true
+homepage = "https://github.com/rust-lang/cargo"
```
Comment (hi-rustin) [inline crates/cargo-util-schemas/Cargo.toml:7]: Off-topic question: Why don't we put the homepage and repository in the `workspace.package`?

**Review by hi-rustin — APPROVED** [review]
Looks good to me. Thanks! 👍

But it seems the third commit message contains a typo. I guess it should be `refactor(sources): Don't impl on schema types`.

File: src/cargo/sources/git/utils.rs [hunk 2]
```diff
@@ -207,56 +207,54 @@ impl GitDatabase {
...
-            GitReference::Branch(s) => {
-                let name = format!("origin/{}", s);
-                let b = repo
-                    .find_branch(&name, git2::BranchType::Remote)
-                    .with_context(|| format!("failed to find branch `{}`", s))?;
-                b.get()
-                    .target()
-                    .ok_or_else(|| anyhow::format_err!("branch `{}` did not have a target", s))?
-            }
+/// Resolves self to an object ID with objects the `repo` currently has.
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:216]:
```suggestion
/// Resolves an [`GitReference`] to an object ID with objects the `repo` currently has.
```

Though personally I prefer an extension trait for grouping these methods.

**Comment by weihanglo** [comment]
> We can always rename and I expect we will when packages-as-namespaces is stabilized

Good point. Naming is a minor issue that might not worth blocking this. I was just trying to avoid anchoring bias. I mean once this is merged, people might be less likely to have opinions on minor details. @epage, do you suggest cancelling this FCP and instead I just r+ this? Totally okay with that.

File: src/cargo/sources/git/utils.rs (hunk 2 above)
Comment (epage) [reply src/cargo/sources/git/utils.rs]: For `PackageIdSpec`, the type is widely used and so having a single trait to pull in and access the functionality seems reasonable.  For this, its pretty limited in where its used (with one exception) which makes me lean the other way.  Or to put this another way, if I were writing this code from scratch, this is likely how I would have done it.

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
People generally agree on the idea so going to merge this now.

@bors r+

**Pull request merged**